lib/pxutil.js

```diff
@@ -267,57 +267,63 @@ function getTokenObject(cookie, delimiter = ':') {
-    const isGraphqlPath = req.path.includes('graphql') || req.originalUrl.includes('graphql');
-    if (!isGraphqlPath || !req.body) {
-        return null;
+function isGraphql(req, config) {
+    if (req.method.toLowerCase() !== 'post') {
+        return false;
     }
 
-    const { body, query } = getGraphqlBodyAndQuery(req);
-    if (!body || !query) {
-        return null;
+    const routes = config['GRAPHQL_ROUTES'];
+    if (!Array.isArray(routes)) {
+        config.logger.error('Invalid configuration px_graphql_routes');
+        return false;
     }
-
-    const operationType = extractGraphqlOperationType(query);
-    const operationName = extractGraphqlOperationName(query, body['operationName']);
-    return new GraphqlData(operationType, operationName);
-}
-
-function getGraphqlBodyAndQuery(req) {
-    let body = {};
-    let query = '';
-
     try {
-        body = typeof req.body === 'string' ? JSON.parse(req.body) : req.body;
-        query = body['query'];
-    } catch (err) {
-        // json parse error
+        return routes.some(r => new RegExp(r).test(req.baseUrl + req.path));
+    } catch (e) {
+        config.logger.error(`Failed to process graphql routes. exception: ${e}`);
+        return false;
     }
-
-    return { body, query };
 }
 
-function extractGraphqlOperationType(query) {
-    const isGraphqlQueryShorthand = query[0] === '{';
-    if (isGraphqlQueryShorthand) {
-        return GraphqlOperationType.QUERY;
+// query: string (not null)
+// output: Record [ OperationName -> OperationType ]
+function parseGraphqlBody(query) {
+    const pattern = /\s*(query|mutation|subscription)\s+(\w+)/gm;
+    let match;
+    const ret = {};
+    while ((match = pattern.exec(query)) !== null) {
+        const operationName = match[2];
+        const operationType = match[1];
+
+        // if two operations has the same name, the query is illegal.
+        if (ret[operationName]) {
+            return null;
+        } else {
+            ret[operationName] = operationType;
+        }
     }
 
-    const queryArray = query.split(/[^A-Za-z0-9_]/);
-    return isValidGraphqlOperationType(queryArray[0]) ? queryArray[0] : GraphqlOperationType.QUERY;
+    return ret;
 }
+// graphqlBodyObject: {query: string?, operationName: string?, variables: any[]?}
+// output: GraphqlData?
+function getGraphqlData(graphqlBodyObject) {
+    if (!graphqlBodyObject || !graphqlBodyObject.query) {
+        return null;
+    }
 
-function extractGraphqlOperationName(query, operationName) {
-    if (operationName) {
-        return operationName;
+    const parsedData = parseGraphqlBody(graphqlBodyObject.query);
+    if (!parsedData) {
+        return null;
     }
 
-    const queryArray = query.split(/[^A-Za-z0-9_]/);
-    return isValidGraphqlOperationType(queryArray[0]) ? queryArray[1] : queryArray[0];
-}
+    const selectedOperationName = graphqlBodyObject['operationName'] ||
+        (Object.keys(parsedData).length === 1 && Object.keys(parsedData)[0]);
```

Comment:
I don't think you need Object.keys(parsedData).length === 1

Reply:
We do, the query is legal when no opeartionName is specified only if there is one operation